### PR TITLE
Deprecate @higherkind & @extension for Either

### PR DESCRIPTION
**Pull request by nomisRev**
- [x] Eq (EqK, EqK2)
- [x] Order
- [x] Hash
- [x] Show
- [x] Semigroup (SemigroupK)
- [x] Monoid
- [x] Functor    
- [x] Apply
- [x] Applicative
- [x] ApplicativeError
- [x] Monad
- [x] MonadError
- [x] Traverse
- [x] BiTraverse
- [x] Foldable
- [x] BiFoldable
- [x] BiFunctor
- [x] BiCrosswalk

Needs to be merged together with:
 - https://github.com/arrow-kt/arrow-incubator/pull/140
 - https://github.com/arrow-kt/arrow-fx/pull/354

### CODE REVIEW

File: arrow-core-data/src/main/kotlin/arrow/core/Either.kt [hunk 1]
```diff
@@ -1041,24 +1162,52 @@ sealed class Either<out A, out B> : EitherOf<A, B> {
     inline fun <L, R> conditionally(test: Boolean, ifFalse: () -> L, ifTrue: () -> R): Either<L, R> =
       if (test) right(ifTrue()) else left(ifFalse())
 
+    @Deprecated("Use the inline version. Hidden for binary compat", level = DeprecationLevel.HIDDEN)
     suspend inline fun <R> catch(f: suspend () -> R): Either<Throwable, R> =
+      catch { f() }
+
+    inline fun <R> catch(f: () -> R): Either<Throwable, R> =
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/Either.kt:1169]: This breaks the API in some places where you pass a `suspend` lambda directly to the `catch` function, I think the Kotlin compiler has the same limitation for suspend method references..

See https://github.com/arrow-kt/arrow-fx/pull/354 as example of this breakage.

**Review by 1Jajen1 — APPROVED** [review]
Looks good to me, thanks! :clap: 

So our approach for now is copy the (deprecated) kapt output, remove kapt, add meta?

**Review by i-walker — APPROVED** [review]
🙌 Thanks Simon

**Comment by 1Jajen1** [comment]
Looking at how https://github.com/arrow-kt/arrow-fx/pull/354 fails this seems to cause some other problems as well :/

**Comment by nomisRev** [comment]
> So our approach for now is copy the (deprecated) kapt output, remove kapt, add meta?

Yes exactly. That's the most graceful way to get rid of the things we want to get rid of for 1.0. Almost all signatures can be preserved, except `traverse` which relies on `Applicative`. So that results in a couple of signatures instead, which need to be differentiated by function name atm due to a bug in the compiler where it cannot differentiate on the return type of the lambda.

**Comment by 1Jajen1** [comment]
> due to a bug in the compiler where it cannot differentiate on the return type of the lambda.

I think this is fixable in some cases with a custom `@jvmName` iirc.

**Comment by nomisRev** [comment]
@1Jajen1 I originally thought so too, but we asked in #compiler on Kotlin Slack and doesn't seem to be the case. See https://youtrack.jetbrains.com/issue/KT-43872

**Pull request merged**